spec/requests/api/connect/v3/systems/products_controller_spec.rb

```diff
@@ -207,90 +207,82 @@
   end
 
   describe '#upgrade' do
+    let(:request) { put url, headers: headers, params: payload }
+    let(:new_product) { FactoryGirl.create(:product, :with_mirrored_repositories) }
+    let(:payload) do
+      {
+        identifier: new_product.identifier,
+        version: new_product.version,
+        arch: new_product.arch
+      }
+    end
+    let(:serialized_json) do
+      V3::ServiceSerializer.new(
+        new_product.service,
+        base_url: URI::HTTP.build({ scheme: response.request.scheme, host: response.request.host }).to_s
+      ).to_json
+    end
+
     it_behaves_like 'products controller action' do
       let(:verb) { 'put' }
     end
 
-    context 'with not activated product' do
-      before { put url, headers: headers, params: payload }
-      subject { response }
+    before do
+      request
+      system.reload
+    end
 
-      let(:product) { FactoryGirl.create(:product, :with_mirrored_repositories) }
-      let(:payload) do
-        {
-          identifier: product.identifier,
-          version: product.version,
-          arch: product.arch
-        }
-      end
+    context 'new product' do
+      subject { response }
 
-      let(:error_response) do
-        {
-          type: 'error',
-          error: "No activation with product '#{product.friendly_name}' was found.",
-          localized_error: "No activation with product '#{product.friendly_name}' was found."
-        }
-      end
+      specify { expect(system.activations.count).to eq(1) }
 
+      its(:code) { is_expected.to eq('201') }
+      its(:body) { is_expected.to eq(serialized_json) }
 
-      its(:code) { is_expected.to eq('422') }
-      its(:body) { is_expected.to eq(error_response.to_json) }
+      it 'activates new product' do
+        expect(system.activations.first.reload.service_id).to equal(new_product.service.id)
+      end
     end
 
-    context 'with activated product' do
-      let(:request) { put url, headers: headers, params: payload }
+    context 'with activated previous product' do
+      subject { response }
```

Comment:
`subject` in all of those examples is the same and can be moved out to describe block.